x-pack/beatless/beater/beatless.go

```diff
@@ -5,68 +5,162 @@
+	if err != nil {
+		return nil, err
 	}
+	ctx, cancel := context.WithCancel(context.Background())
 
 	bt := &Beatless{
-		done:   make(chan struct{}),
-		config: c,
-		log:    logp.NewLogger("beatless"),
+		ctx:      ctx,
+		cancel:   cancel,
+		log:      logp.NewLogger("beatless"),
+		Provider: provider,
+		Config:   c,
 	}
 	return bt, nil
 }
 
 // Run starts beatless.
 func (bt *Beatless) Run(b *beat.Beat) error {
+	defer bt.cancel()
 	bt.log.Info("beatless is running")
 	defer bt.log.Info("beatless stopped running")
 
-	client, err := b.Publisher.Connect()
+	manager, err := licenser.Create(&b.Config.Output, refreshDelay, graceDelay)
 	if err != nil {
+		return errors.Wrap(err, "could not create the license manager")
+	}
+	manager.Start()
+	defer manager.Stop()
+
+	// Wait until we receive the initial license.
+	if err := licenser.WaitForLicense(bt.ctx, bt.log, manager, checkLicense); err != nil {
 		return err
 	}
-	defer client.Close()
-
-	// NOTE: Do not review below, this is the minimal to have a working PR.
-	bus := bus.New(client)
-	// TODO: noop
-	bus.Listen()
-
-	// Stop until we are tell to shutdown.
-	// TODO this is where the events catcher starts.
-	select {
-	case <-bt.done:
-		// Stop catching events.
+
+	clientFactory := makeClientFactory(bt.log, manager, b.Publisher)
+
+	enabledFunctions := bt.enabledFunctions()
+	bt.log.Infof("enabled functions: %s", strings.Join(enabledFunctions, ", "))
+	// Create a client per function and wrap them into a runnable function by the coordinator.
+	functions, err := bt.Provider.CreateFunctions(clientFactory, enabledFunctions)
+	if err != nil {
+		return fmt.Errorf("error when creating the functions, error: %v", err)
+	}
+
+	// manages the goroutine related to the function handlers, if an error occurs and its not handled
+	// by the function itself, it will reach the coordinator, we log the error and shutdown beats.
+	// When an error reach the coordinator we assume that we cannot recover from it and we initiate
+	// a shutdown and return an aggregated errors.
+	coordinator := core.NewCoordinator(logp.NewLogger("coordinator"), functions...)
+	err = coordinator.Run(bt.ctx)
+	if err != nil {
+		return err
 	}
 	return nil
 }
 
+func (bt *Beatless) enabledFunctions() (values []string) {
+	raw, found := os.LookupEnv("ENABLED_FUNCTIONS")
+	if !found {
+		return values
+	}
+	return strings.Split(raw, ",")
+}
+
 // Stop stops beatless.
 func (bt *Beatless) Stop() {
 	bt.log.Info("beatless is stopping")
 	defer bt.log.Info("beatless is stopped")
-	close(bt.done)
+	bt.cancel()
+}
+
+func makeClientFactory(log *logp.Logger, manager *licenser.Manager, pipeline beat.Pipeline) func(*common.Config) (core.Client, error) {
+	// Each function has his own client to the publisher pipeline,
+	// publish operation will block the calling thread, when the method unwrap we have received the
+	// ACK for the batch.
+	clientFactory := func(cfg *common.Config) (core.Client, error) {
```

Comment:
as makeClientFactory is the constructor now, just use `return func(cfg *common.Config) (core.Client, error)`. This makes it immediately clear that clientFactory will not be used inside this function.